crates/wiggle/src/borrow.rs

```diff
@@ -1,259 +1,113 @@
-    /// to be safe to recursively call into a WebAssembly module, or to
-    /// manipulate the WebAssembly memory by any other means.
-    pub fn has_outstanding_borrows(&self) -> bool {
-        self.bc.lock().unwrap().has_outstanding_borrows()
-    }
     pub fn shared_borrow(&self, r: Region) -> Result<BorrowHandle, GuestError> {
-        self.bc.lock().unwrap().shared_borrow(r)
+        match self.state.load(Relaxed) {
+            n if n >= u32::MAX - 1 => Err(GuestError::PtrBorrowed(r)),
+            n => {
+                self.state.store(n + 1, Relaxed);
+                Ok(BorrowHandle { _priv: () })
+            }
+        }
     }
     pub fn mut_borrow(&self, r: Region) -> Result<BorrowHandle, GuestError> {
-        self.bc.lock().unwrap().mut_borrow(r)
-    }
-    pub fn shared_unborrow(&self, h: BorrowHandle) {
-        self.bc.lock().unwrap().shared_unborrow(h)
-    }
-    pub fn mut_unborrow(&self, h: BorrowHandle) {
-        self.bc.lock().unwrap().mut_unborrow(h)
-    }
-    pub fn is_shared_borrowed(&self, r: Region) -> bool {
-        self.bc.lock().unwrap().is_shared_borrowed(r)
-    }
-    pub fn is_mut_borrowed(&self, r: Region) -> bool {
-        self.bc.lock().unwrap().is_mut_borrowed(r)
-    }
-}
-
-#[derive(Debug)]
-/// This is a pretty naive way to account for borrows. This datastructure
-/// could be made a lot more efficient with some effort.
-struct InnerBorrowChecker {
-    /// Maps from handle to region borrowed. A HashMap is probably not ideal
-    /// for this but it works. It would be more efficient if we could
-    /// check `is_borrowed` without an O(n) iteration, by organizing borrows
-    /// by an ordering of Region.
-    shared_borrows: HashMap<BorrowHandle, Region>,
-    mut_borrows: HashMap<BorrowHandle, Region>,
-    /// Handle to give out for the next borrow. This is the bare minimum of
-    /// bookkeeping of free handles, and in a pathological case we could run
-    /// out, hence [`GuestError::BorrowCheckerOutOfHandles`]
-    next_handle: BorrowHandle,
-}
-
-impl InnerBorrowChecker {
-    fn new() -> Self {
-        InnerBorrowChecker {
-            shared_borrows: HashMap::new(),
-            mut_borrows: HashMap::new(),
-            next_handle: BorrowHandle(0),
+        match self.state.load(Relaxed) {
+            0 => {
+                self.state.store(u32::MAX, Relaxed);
+                Ok(BorrowHandle { _priv: () })
+            }
+            _ => Err(GuestError::PtrBorrowed(r)),
         }
     }
-
-    fn has_outstanding_borrows(&self) -> bool {
-        !(self.shared_borrows.is_empty() && self.mut_borrows.is_empty())
+    pub fn shared_unborrow(&self, _: BorrowHandle) {
+        let prev = self.state.load(Relaxed);
+        debug_assert!(prev > 0);
+        self.state.store(prev - 1, Relaxed);
     }
-
-    fn is_shared_borrowed(&self, r: Region) -> bool {
-        self.shared_borrows.values().any(|b| b.overlaps(r))
+    pub fn mut_unborrow(&self, _: BorrowHandle) {
```

Comment:
Should there be different handle types for mut vs shared borrows?

Reply:
In theory yeah, but I'm not too keen to do too much refactoring here because it's all basically a dated library on life support now. These are all ignored anyway so I should probably refactor to remove them entirely, but I'll leave that as a future TODO if necessary